internal/argocd/revision.go

```diff
@@ -2,107 +2,224 @@ package argocd
+			sourceLogger.Error(err, "Error finding chart from repo")
+			return "", fmt.Errorf("error chart from repo %q: %w", s.RepoURL, err)
 		}
 		if chart == nil {
+			sourceLogger.Debug("Chart not found")
 			return "", nil
 		}
 		return chart.Version, nil
-	case app.Spec.Source.RepoURL != "":
-		// This source points to a Git repository.
-
-		// If there is a source update that targets app.Spec.Source, it might
-		// have its own ideas about the desired revision.
-		var targetPromoMechanism any
-		for i := range update.SourceUpdates {
-			sourceUpdate := &update.SourceUpdates[i]
-			if sourceUpdate.RepoURL == app.Spec.Source.RepoURL {
-				targetPromoMechanism = sourceUpdate
-				break
-			}
-		}
-		if targetPromoMechanism == nil {
-			targetPromoMechanism = update
-		}
-		desiredOrigin := freight.GetDesiredOrigin(stage, targetPromoMechanism)
+
+	case s.RepoURL != "":
+		sourceLogger.Debug("Application source is a Git repository")
 		commit, err := freight.FindCommit(
 			ctx,
 			cl,
 			stage,
 			desiredOrigin,
 			frght,
-			app.Spec.Source.RepoURL,
+			s.RepoURL,
 		)
 		if err != nil {
+			sourceLogger.Error(err, "Error finding commit from repo")
 			return "",
-				fmt.Errorf("error finding commit from repo %q: %w", app.Spec.Source.RepoURL, err)
+				fmt.Errorf("error finding commit from repo %q: %w", s.RepoURL, err)
 		}
 		if commit == nil {
+			sourceLogger.Debug("Commit not found")
 			return "", nil
 		}
 		if commit.HealthCheckCommit != "" {
 			return commit.HealthCheckCommit, nil
 		}
 		return commit.ID, nil
 	}
-	// If we end up here, no desired revision was found.
+
+	sourceLogger.Debug("Could not determine desired revision for application from this source.")
 	return "", nil
 }
+
+func getDesiredOriginForSource(
+	ctx context.Context,
+	stage *kargoapi.Stage,
+	update *kargoapi.ArgoCDAppUpdate,
+	s *argocd.ApplicationSource,
+) *kargoapi.FreightOrigin {
+	sourceLogger := logging.LoggerFromContext(ctx).WithValues("source", s)
+	sourceLogger.Debug("Resolving origin for application source")
+
+	// If there is a source update that targets app.Spec.Source, it might
+	// have its own ideas about the desired revision.
+	// Default to the ArgoCDAppUpdate if no source update targets app.Spec.Source.
+	var targetPromoMechanism any
+
+	for i := range update.SourceUpdates {
+		sourceUpdate := &update.SourceUpdates[i]
+		sourceLogger.Trace("Checking source update", "sourceUpdate", sourceUpdate)
+		if sourceUpdate.RepoURL != s.RepoURL {
+			sourceLogger.Debug("Source update does not match application source, skipping", "sourceUpdate", sourceUpdate)
+			continue
+		}
+
+		if s.Chart != "" && sourceUpdate.Chart != s.Chart {
+			sourceLogger.Debug("Source update chart does not match application source chart, skipping",
+				"sourceUpdate", sourceUpdate)
+			continue
+		}
+
+		sourceLogger.Debug("Source update matching application source found", "sourceUpdate", sourceUpdate)
+		targetPromoMechanism = sourceUpdate
+		break
```

Comment:
This logic isn't wrong, but it's a bit difficult to follow. Instead of two cases where you want to continue and then dealing with the case where you found what you were looking for as a default, you could write this as just a single case of what to do when you _have_ found what you're looking for. It's fewer cognitive hoops to jump through to arrive at an understanding of what it's doing.